File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.py
# 1932
import sys
input = sys.stdin.readline
n = int(input())
arr = []
for _ in range(n):
    arr.append(list(map(int, input().split())))

w = 2
for i in range(1, n):
    for j in range(w):
        if j == 0: 
            arr[i][j] = arr[i-1][0] + arr[i][j]
        elif i == j:
            arr[i][j] = arr[i-1][j-1] + arr[i][j]
        else:
            arr[i][j] = max(arr[i-1][j-1], arr[i-1][j]) + arr[i][j]
    w += 1

print(max(arr[n-1]))

File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        int n = Integer.parseInt(br.readLine());
        List<Integer>[] arr = new ArrayList[n+1];
        for (int i = 0; i < n+1; i++) {
            arr[i] = new ArrayList<>();
        }
        for (int i = 1; i < n+1; i++) {
            st = new StringTokenizer(br.readLine());
            while (st.hasMoreTokens()) {
                arr[i].add(Integer.parseInt(st.nextToken()));
            }
        }

        for (int i = 2; i < n+1; i++) {
            for (int j = 0; j < i; j++) {
                if (j == 0) {
                    arr[i].set(j, arr[i-1].get(0) + arr[i].get(0));
                } else if (j == i-1) {
                    arr[i].set(j, arr[i-1].get(j-1) + arr[i].get(j));
                } else {
                    arr[i].set(j, Math.max(arr[i-1].get(j-1), arr[i-1].get(j)) + arr[i].get(j));
                }
            }
        }

        int res = 0;
        for (int i = 0; i < n; i++) {
            res = Math.max(res, arr[n].get(i));
        }
        System.out.println(res);
    }
}
